packages/x-data-grid/src/components/columnSelection/GridCellCheckboxRenderer.tsx

```diff
@@ -45,7 +50,23 @@ const GridCellCheckboxForwardRef = forwardRef<HTMLInputElement, GridRenderCellPa
     const ownerState = { classes: rootProps.classes };
     const classes = useUtilityClasses(ownerState);
 
+    const { isIndeterminate, isChecked, isSelectable } = useGridSelector(
+      apiRef,
+      checkboxPropsSelector,
+      {
+        groupId: id,
+        autoSelectParents: rootProps.rowSelectionPropagation?.parents ?? false,
+        isRowSelectable: rootProps.isRowSelectable,
+        columns: rootProps.columns ?? EMPTY_ARRAY,
```

Comment:
`rootProps.columns` gets outdated during the lifecycle of a Data Grid; we should be using selectors instead (e.g., `gridColumnDefinitionsSelector`) to access columns.